chore(update): config docs w/ superflags (#113)
dgraph-io/dgraph-docs: content/deploy/config.md

## Command line flags

> Dgraph has *global flags* that apply to all subcommands and flags specific to a subcommand.

Reviewer: I'd love to use the term "command" whenever possible, whether we are talking about the root `dgraph` command, subcommands of `dgraph` like `alpha` and `acl`, or the next level down when it exists. We should ideally use the term "subcommand" only when we need to disambiguate and clarify command relationships, as in "subcommands of `dgraph` (I'm happy to do a global edit for this) :).

So, if we just use the term "command" when talking about compound commands that use flags, we will simplify our docs. We can definitely use "subcommand" to disambiguate in a context like "subcommand of Dgraph".

Author: I didn't want to create terminology that would be inconsistent with POSIX shell, Unix, Linux, and 50+ years of precedent with Bourne shell.

Reviewer: OK, but "command" is what is used in the command-line help.

> Starting with release `v21.03.0` comes the introduction of superflags.  Flags of a similar category have been consolidated into a single superflag followed by string that is semicolon-delimited list of option value pairs.  So the general syntax for flags will be: `--<flagname> option-a=value;option-b=value`

Reviewer: I'd rather avoid adding this term to the docs, at least as a preferred term. "Compound flags" would work if we must have a term for these, otherwise we can just rewrite to describe the new flags (see next comment).

Reviewer: Also, "release 21.03.0" can be in plain text.

Author: This is the term used throughout the code, and used by Dgraph?!?

> Starting with release `v21.03.0` comes the introduction of superflags.  Flags of a similar category have been consolidated into a single superflag followed by string that is semicolon-delimited list of option value pairs.  So the general syntax for flags will be: `--<flagname> option-a=value;option-b=value`

Reviewer: ```suggestion
Some flags have been deprecated and replaced in release 21.03, and flags for several commands (`alpha`, `backup`, `bulk`,`debug`, `live`, and `zero`) now have compound flags. Compound flags contain one or more options that let you define multiple settings in a semicolon-delimited list that's encapsulated in double-quotes. The general syntax for compound flags is as follows: `--<flagname> "option-a=value-a; option-b=value-b"`.

Author: I will use the snippet, but the _encapsulated in double-quotes_ is functionally incorrect.  Double quotes are only needed if spaces are in the argument, and not sure if we parse spaces in the string.

Author: The version string would be inconsistent with releases, docker image releases, and documentation releases.  I am fine if we still want to use the other way.

> Below is an example of using command line flags with `dgraph alpha`.

Reviewer: ```suggestion
The following example shows how to use compound flags when running the `dgraph alpha` command.
```

```bash
dgraph alpha --my=alpha.example.com:7080 --zero=zero.example.com:5080 \
  --badger "compression=zstd:1" \
  --block_rate "10" \
  --trace "jaeger=http://jaeger:14268" \
  --tls "ca-cert=/dgraph/tls/ca.crt;client-auth-type=REQUIREANDVERIFY;server-cert=/dgraph/tls/node.crt;server-key=/dgraph/tls/node.key;use-system-ca=true;internal-port=true;client-cert=/dgraph/tls/client.dgraphuser.crt;client-key=/dgraph/tls/client.dgraphuser.key"
  --security "whitelist=10.0.0.0/8,172.0.0.0/8,192.168.0.0/16"
```

## Environment variables

> The environment variable names mirror the flag names as seen in the `--help` output. They are the concatenation of `DGRAPH`, the subcommand invoked

Reviewer: ```suggestion
Dgraph environment variable names mirror the flag names shown in the Dgraph CLI `--help` output. These environment variable names are formed by concatenating `DGRAPH`, the subcommand invoked
```

(`ALPHA`, `ZERO`, `LIVE`, or `BULK`), and then the name of the flag (in uppercase). For example, instead of using `dgraph alpha --block_rate 10`, you

> could use `DGRAPH_ALPHA_BLOCK_RATE=10 dgraph alpha`.

Reviewer: Suggestion:
"
For example, instead of using `dgraph alpha --block_rate 10`, you
could use `DGRAPH_ALPHA_BLOCK_RATE=10 dgraph alpha`. 
"

Changes to:

"
For example, instead running a command like `dgraph alpha --block_rate 10`, you could set the following environment variable: `DGRAPH_ALPHA_BLOCK_RATE=10 dgraph alpha`.
"

For an environment variable, the `--<flagname> option-a=value;option-b=value` can be represented as `<FLAGNAME>="option-a=value;option-b=value"`.

Below is an example of environment variables for `dgraph alpha`:

```bash
DGRAPH_ALPHA_BADGER="compression=zstd:1"
DGRAPH_ALPHA_BLOCK_RATE="10"
DGRAPH_ALPHA_TRACE="jaeger=http://jaeger:14268"
DGRAPH_ALPHA_TLS="ca-cert=/dgraph/tls/ca.crt;client-auth-type=REQUIREANDVERIFY;server-cert=/dgraph/tls/node.crt;server-key=/dgraph/tls/node.key;use-system-ca=true;internal-port=true;client-cert=/dgraph/tls/client.dgraphuser.crt;client-key=/dgraph/tls/client.dgraphuser.key"
DGRAPH_ALPHA_SECURITY="whitelist=10.0.0.0/8,172.0.0.0/8,192.168.0.0/16"
```

## Configuration file

A configuration file can be specified using the `--config` flag, or an environment variable, such as `dgraph alpha --config my_config.json` or `DGRAPH_ALPHA_CONFIG=my_config.json dgraph alpha`.

> Dgraph supports configuration file formats that it detects based on file extensions ([`.json`](https://www.json.org/json-en.html), [`.yml`](https://yaml.org/) or [`.yaml`](https://yaml.org/)).  In these formats, the superflag will be used as a key that points to a hash.  The hash itself consists of `key: value` pairs corresponds to the superflag list of `option=value` pairs.

Reviewer: ```suggestion
Dgraph supports configuration file formats that it detects based on file extensions ([`.json`](https://www.json.org/json-en.html), [`.yml`](https://yaml.org/) or [`.yaml`](https://yaml.org/)).  In these files, the name of the compound flag is used as a key that points to a hash. The hash consists of `key: value` pairs that correspond to the compound flag's list of `option=value` pairs.
```

{{% notice "note" %}}
The formats [`.toml`](https://toml.io/en/), [`.hcl`](https://github.com/hashicorp/hcl), and [`.properties`](https://en.wikipedia.org/wiki/.properties) are not supported in release `v21.03.0`.
{{% /notice %}}

{{% notice "tip" %}}
When representing the superflag options in the hash, you can use either *kebab-case* or *snake_case* for names of the keys.
{{% /notice %}}

### JSON config file

> For JSON, the `--<flagname> option-a=value;option-b=value` can be represented as:

Reviewer: ```suggestion
In JSON, `--<flagname> option-a=value;option-b=value` can be represented as follows:
```

```json
{
  "<flagname>": {
    "option-a": "value",
    "opton-b": "value"
  }
}
```

> Example JSON config file (`config.json`) using *kebab-case*:

Reviewer: ```suggestion
The following example JSON config file (`config.json`) uses *kebab-case*:
```

```json
{
  "badger": { "compression": "zstd:1" },
  "trace": { "jaeger": "http://jaeger:14268" },
  "security": { "whitelist": "10.0.0.0/8,172.0.0.0/8,192.168.0.0/16" },
  "tls": {
    "ca-cert": "/dgraph/tls/ca.crt",
    "client-auth-type": "REQUIREANDVERIFY",
    "server-cert": "/dgraph/tls/node.crt",
    "server-key": "/dgraph/tls/node.key",
    "use-system-ca": true,
    "internal-port": true,
    "client-cert": "/dgraph/tls/client.dgraphuser.crt",
    "client-key": "/dgraph/tls/client.dgraphuser.key"
  }
}
```

> Example JSON config file (`config.json`) using *snake_case*:

Reviewer: ```suggestion
The following example JSON config file (`config.json`) uses *snake_case*:
```

```json
{
  "badger": { "compression": "zstd:1" },
  "trace": { "jaeger": "http://jaeger:14268" },
  "security": { "whitelist": "10.0.0.0/8,172.0.0.0/8,192.168.0.0/16" },
  "tls": {
    "ca_cert": "/dgraph/tls/ca.crt",
    "client_auth_type": "REQUIREANDVERIFY",
    "server_cert": "/dgraph/tls/node.crt",
    "server_key": "/dgraph/tls/node.key",
    "use_system_ca": true,
    "internal_port": true,
    "client_cert": "/dgraph/tls/client.dgraphuser.crt",
    "client_key": "/dgraph/tls/client.dgraphuser.key"
  }
}
```

### YAML config file

> For YAML, the `--<flagname> option-a=value;option-b=value` can be represented as:

Reviewer: For YAML, the -> In YAML,
